File: 31-ways-to-climb-stairs/JavaScript/solution1.js
// By using recursion:

// Time complexity: O(m^n)
// Space complexity: O(n)

function waysToClimb(n, possibleSteps) {
    if (n == 0) return 1;
    let nbWays = 0;
    for (let steps of possibleSteps)
        if (n - steps >= 0) nbWays += waysToClimb(n - steps, possibleSteps);
    return nbWays;
}

waysToClimb(7, {2, 3, 4});

// output 5
// 2 2 3
// 2 3 2
// 3 2 2
// 3 4
// 4 3

File: 19-paths-in-matrix/Python/solution1.py
# By using recursion:

# Time complexity: O(2^(n*m))
# Space complexity: O(n + m)

def paths(matrix, i = 0, j = 0):
  n = len(matrix)
  m = len(matrix[0])
  if i > n-1 or j > m-1 or matrix[i][j] == 1:
    return 0
  elif i == n-1 and j == m-1:
    return 1
  else:
    return paths(matrix, i+1, j) + paths(matrix, i, j+1)


